utilities/PerformanceTesting/GetAverages.py

```diff
@@ -0,0 +1,147 @@
+import argparse
+import json
+import sys
+from argparse import RawTextHelpFormatter
+
+
+def getFilePaths(file_prefix, num_files):
+    """
+    Generates the json file paths of the format <file_prefix>idx.json
+    where idx is an integer between 0 and num_files-1
+
+    :param file_prefix: The first part of the file path
+    that all files to be averaged have in common
+    :param num_files: The number of files to be averaged
+
+    :return: A list of files to be averaged
+    """
+    files = []
+
+    for i in range(num_files):
+        files.append(f"{file_prefix}{i}.json")
+
+    return files
+
+
+def getJsons(file_paths):
+    """
+    Reads the json files to be averaged and returns the json contents
+
+    :param file_paths: A list of file paths
+
+    :return: A list of jsons
+    """
+    jsons = []
+
+    for file_path in file_paths:
+        with open(file_path) as f:
+            jsons.append(json.load(f))
+
+    return jsons
+
+
+def findAverageForMetric(jsons, metric_group, metric):
+    """
+    Finds the average value of a metric over some runs
+
+    :param jsons: A list of jsons containing metrics
+    :param metric_group: The metric group that the metric belongs to
+    :param metric: The metric that is being averaged
+
+    :return: Returns a tuple. The first element is the average or 0
+    if it is not possible to calculate the average.
+    The second element is a boolean specifying if the average was
+    successfully computed.
+    """
+    total = 0
+    count = 0
+
+    for j in jsons:
+        if metric not in j[metric_group]:
+            err_msg = "Missing metric: " + metric
+            raise KeyError(err_msg)
+
+        try:
+            value = j[metric_group][metric]['value']
+            total += float(value)
+            count += 1.0
+        except (ValueError, TypeError):
+            continue
+
+    if count > 0:
+        return total / float(count), True
+    else:
+        return 0, False
+
+
+def calcAverage(file_prefix, num_files):
+    """
+    Computes the average results over multiple runs
+
+    :param file_prefix: The first part of the file path
+    that all files to be averaged have in common
+    :param num_files: The number of files to be averaged
+
+    :return: A json with the average of the results of the different runs
+    """
+
+    files = getFilePaths(file_prefix, num_files)
+
+    jsons = getJsons(files)
+    ave_json = jsons[0]
+
+    for metric_group in jsons[0]:
+        for metric in jsons[0][metric_group]:
+            ave, valid = findAverageForMetric(jsons, metric_group, metric)
```

Comment:
I am not sure how else I can do this. I need to pass the keys to findAverageForMetric so I can find the entry in the other jsons.

Reply:
I think we can try to think about it from a different angle.

Instead of grabbing each metric individually and iterating over the dictionaries, iterate over all dictionaries and add all metrics to a separate dictionary that holds the cumulative sum of them, once you are done iterating over the dictionaries just divide all metrics by the total amount of entries added. You might need to keep a separate list of metrics that need to be part of a dictionary and verify every entry has all of those before adding the metrics, but that should be simple enough IMO. Does it make sense to try something like this?

Reply:
That does make sense. My concern is that currently this script does not fail if an entry is null, or otherwise not a float. It keeps a counter of the number of valid entries that it sees. The motivation for this is that the metric "Average number of incoming L4_PROTOCOL_UDP network flows over pods" is null in every case that I have seen, but I don't know that this will always be the case. So I wanted the code to be able to handle that case. So I would either need to keep a counter within the json, have the script exit when it encounters a null, or assign a null to the final result if one of the entries being averaged over is a null. I am leaning towards the last option.